core/state/state_witness.go

```diff
@@ -1,36 +1,133 @@
 package state
 
 import (
+	"bytes"
 	"fmt"
-
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"os"
 )
 
-type witness struct {
-	root   common.Hash
-	lists  []map[string][]byte
-	owners []common.Hash
+type Witness struct {
+	blockHashes map[uint64]common.Hash
+	codes       map[common.Hash]Code
+	root        common.Hash
+	lists       map[common.Hash]map[string][]byte
 }
 
-func newWitness(originalRoot common.Hash) *witness {
-	return &witness{root: originalRoot}
+func (w *Witness) EncodeRLP(b *types.Block) []byte {
+	buf := new(bytes.Buffer)
+	eb := rlp.NewEncoderBuffer(buf)
+
+	var owners []common.Hash
+	var allPaths [][]string
+	var allNodes [][][]byte
+	var blockNums []uint64
+	var blockHashes []common.Hash
+	var codes []Code
+	var codeHashes []common.Hash
+
+	for owner, nodeMap := range w.lists {
+		owners = append(owners, owner)
+		var ownerPaths []string
+		var ownerNodes [][]byte
+
+		for path, node := range nodeMap {
+			ownerPaths = append(ownerPaths, path)
+			ownerNodes = append(ownerNodes, node)
+		}
+		allPaths = append(allPaths, ownerPaths)
+		allNodes = append(allNodes, ownerNodes)
+	}
+
+	for codeHash, code := range w.codes {
+		codeHashes = append(codeHashes, codeHash)
+		codes = append(codes, code)
+	}
```

Comment:
I don't think codehashes can be "decoupled". Code-hash is part of the account leaf. In order to capture the codehash, we need to supply the entire account leaf, along with proofs. 

Same with blockhashes, for that matter. If we want to be strict, we need to provide a chain of headers going back long enough to cover all used hashes.

Reply:
> I don't think codehashes can be "decoupled". Code-hash is part of the account leaf. In order to capture the codehash, we need to supply the entire account leaf, along with proofs.

It could be useful to explicitly include code-hash paired with the code it corresponds  to because without it, the verifier would have to compute the hash of each received code during validation.

> Same with blockhashes, for that matter. If we want to be strict, we need to provide a chain of headers going back long enough to cover all used hashes.

Yeah we could potentially do this.  For now it's easier to include them without proving their validity.

Reply:
Maybe. But I thought you added them 'standalone' on e.g. `EXTCODEHASH`. If so, that would be wrong. In that case, we need the account leaf.

Reply:
Updated extcodehash to also fetch the account leaf.